File: crates/ui/src/container_icon.rs
use gpui::*;

use crate::{Icon, IconName};
use crate::theme::Theme;

/// A known container technology with its SVG icon path.
pub struct Technology {
    pub name: &'static str,
    pub svg_path: &'static str,
}

/// Attempt to detect a known technology from a container image string.
///
/// Extracts the base image name (last path segment, minus tag/digest) and
/// matches it against a curated list of devicon-supported technologies.
pub fn detect_technology(image: &str) -> Option<&'static Technology> {
    // Extract base name: "registry.io/org/nginx:1.25" → "nginx"
    let base = image.rsplit('/').next().unwrap_or(image);
    let name = base.split(':').next().unwrap_or(base);
    let name = name.split('@').next().unwrap_or(name); // handle digest refs

    lookup_technology(name)
}

fn lookup_technology(name: &str) -> Option<&'static Technology> {
    let lower = name.to_ascii_lowercase();
    TECHNOLOGIES.iter().find(|t| {
        t.name == lower
            || lower.starts_with(t.name)
            || lower.contains(t.name)
    })
}

/// Render a container icon for the given image. Shows a technology-specific SVG
/// when the image matches a known technology, otherwise falls back to the generic Box icon.
/// Uses `img()` instead of `svg()` to preserve the original multi-color SVG rendering.
pub fn container_icon(image: &str, theme: &Theme) -> AnyElement {
    let colors = &theme.colors;

    if let Some(tech) = detect_technology(image) {
        div()
            .flex_shrink_0()
            .size(px(36.0))
            .rounded(theme.border_radius_md)
            .bg(colors.surface)
            .border_1()
            .border_color(colors.border)
            .flex()
            .items_center()
            .justify_center()
            .child(img(tech.svg_path).size(px(20.0)))
            .into_any_element()
    } else {
        div()
            .flex_shrink_0()
            .size(px(36.0))
            .rounded(theme.border_radius_md)
            .bg(colors.primary)
            .flex()
            .items_center()
            .justify_center()
            .child(
                Icon::new(IconName::Box)
                    .size(px(18.0))
                    .color(colors.background),
            )
            .into_any_element()
    }
}

/// Render a small container icon (22px) for use in compact layouts like table rows.
/// Uses a light surface background with the original multi-color SVG.
pub fn container_icon_small(image: &str, theme: &Theme) -> AnyElement {
    let colors = &theme.colors;

    if let Some(tech) = detect_technology(image) {
        div()
            .flex_shrink_0()
            .size(px(22.0))
            .rounded(theme.border_radius_sm)
            .bg(colors.surface)
            .border_1()
            .border_color(colors.border)
            .flex()
            .items_center()
            .justify_center()
            .child(img(tech.svg_path).size(px(14.0)))
            .into_any_element()
    } else {
        div()
            .flex_shrink_0()
            .size(px(22.0))
            .rounded(theme.border_radius_sm)
            .bg(colors.surface)
            .border_1()
            .border_color(colors.border)
            .flex()
            .items_center()
            .justify_center()
            .child(
                Icon::new(IconName::Box)
                    .size(px(12.0))
                    .color(colors.text_muted),
            )
            .into_any_element()
    }
}

// Technology database — SVG icon paths (loaded from assets)
static TECHNOLOGIES: &[Technology] = &[
    Technology { name: "nginx", svg_path: "icons/tech/nginx.svg" },
    Technology { name: "redis", svg_path: "icons/tech/redis.svg" },
    Technology { name: "postgres", svg_path: "icons/tech/postgresql.svg" },
    Technology { name: "mysql", svg_path: "icons/tech/mysql.svg" },
    Technology { name: "mongo", svg_path: "icons/tech/mongodb.svg" },
    Technology { name: "python", svg_path: "icons/tech/python.svg" },
    Technology { name: "node", svg_path: "icons/tech/nodejs.svg" },
    Technology { name: "golang", svg_path: "icons/tech/go.svg" },
    Technology { name: "go", svg_path: "icons/tech/go.svg" },
    Technology { name: "docker", svg_path: "icons/tech/docker.svg" },
    Technology { name: "kubernetes", svg_path: "icons/tech/kubernetes.svg" },
    Technology { name: "k8s", svg_path: "icons/tech/kubernetes.svg" },
    Technology { name: "java", svg_path: "icons/tech/java.svg" },
    Technology { name: "openjdk", svg_path: "icons/tech/java.svg" },
    Technology { name: "eclipse-temurin", svg_path: "icons/tech/java.svg" },
    Technology { name: "rust", svg_path: "icons/tech/rust.svg" },
    Technology { name: "php", svg_path: "icons/tech/php.svg" },
    Technology { name: "fpm", svg_path: "icons/tech/php.svg" },
    Technology { name: "ruby", svg_path: "icons/tech/ruby.svg" },
    Technology { name: "react", svg_path: "icons/tech/react.svg" },
    Technology { name: "typescript", svg_path: "icons/tech/typescript.svg" },
    Technology { name: "apache", svg_path: "icons/tech/apache.svg" },
    Technology { name: "httpd", svg_path: "icons/tech/apache.svg" },
    Technology { name: "tomcat", svg_path: "icons/tech/tomcat.svg" },
    Technology { name: "rabbitmq", svg_path: "icons/tech/rabbitmq.svg" },
    Technology { name: "kafka", svg_path: "icons/tech/kafka.svg" },
    Technology { name: "elasticsearch", svg_path: "icons/tech/elasticsearch.svg" },
    Technology { name: "elastic", svg_path: "icons/tech/elasticsearch.svg" },
    Technology { name: "grafana", svg_path: "icons/tech/grafana.svg" },
    Technology { name: "prometheus", svg_path: "icons/tech/prometheus.svg" },
    Technology { name: "jenkins", svg_path: "icons/tech/jenkins.svg" },
    Technology { name: "terraform", svg_path: "icons/tech/terraform.svg" },
    Technology { name: "consul", svg_path: "icons/tech/consul.svg" },
    Technology { name: "cassandra", svg_path: "icons/tech/cassandra.svg" },
    Technology { name: "mariadb", svg_path: "icons/tech/mariadb.svg" },
];

File: src/lib/utils/container-icon.ts
// Maps container image names to devicon icon identifiers
// Uses https://cdn.jsdelivr.net/gh/devicons/devicon@latest/icons/{name}/{name}-{variant}.svg

const IMAGE_TO_DEVICON: Record<string, string> = {
  // Databases
  postgres: "postgresql",
  postgresql: "postgresql",
  mysql: "mysql",
  mariadb: "mariadb",
  mongo: "mongodb",
  mongodb: "mongodb",
  redis: "redis",
  cassandra: "cassandra",
  couchdb: "couchdb",
  neo4j: "neo4j",
  influxdb: "influxdb",
  cockroachdb: "cockroachdb",

  // Web servers & proxies
  nginx: "nginx",
  apache: "apache",
  httpd: "apache",
  traefik: "traefikproxy",
  envoy: "envoy",
  haproxy: "haproxy",
  caddy: "caddy",

  // Languages & runtimes
  node: "nodejs",
  nodejs: "nodejs",
  python: "python",
  golang: "go",
  go: "go",
  ruby: "ruby",
  php: "php",
  java: "java",
  openjdk: "java",
  eclipse_temurin: "java",
  "eclipse-temurin": "java",
  rust: "rust",
  dotnet: "dotnetcore",
  "mcr.microsoft.com/dotnet": "dotnetcore",
  swift: "swift",
  perl: "perl",
  elixir: "elixir",
  erlang: "erlang",
  scala: "scala",
  kotlin: "kotlin",
  clojure: "clojure",
  haskell: "haskell",
  lua: "lua",
  r: "r",
  dart: "dart",
  deno: "denojs",
  bun: "bun",

  // Frameworks
  django: "django",
  flask: "flask",
  fastapi: "fastapi",
  rails: "rails",
  spring: "spring",
  nextjs: "nextjs",
  nuxt: "nuxtjs",
  laravel: "laravel",
  express: "express",
  nestjs: "nestjs",
  svelte: "svelte",
  react: "react",
  vue: "vuejs",
  angular: "angular",

  // Infrastructure & DevOps
  jenkins: "jenkins",
  gitlab: "gitlab",
  grafana: "grafana",
  prometheus: "prometheus",
  elasticsearch: "elasticsearch",
  kibana: "kibana",
  logstash: "logstash",
  fluentd: "fluentd",
  rabbitmq: "rabbitmq",
  kafka: "apachekafka",
  zookeeper: "apache",
  vault: "vault",
  consul: "consul",
  terraform: "terraform",
  ansible: "ansible",
  puppet: "puppet",
  argocd: "argocd",
  argo: "argocd",

  // Container & orchestration
  docker: "docker",
  podman: "podman",
  kubernetes: "kubernetes",

  // Cloud
  amazonlinux: "amazonwebservices",
  "gcr.io": "googlecloud",
  "mcr.microsoft.com": "azure",

  // OS base images
  ubuntu: "ubuntu",
  debian: "debian",
  alpine: "alpinelinux",
  centos: "centos",
  fedora: "fedora",
  archlinux: "archlinux",
  opensuse: "opensuse",

  // Other
  wordpress: "wordpress",
  drupal: "drupal",
  joomla: "joomla",
  tomcat: "tomcat",
  jetty: "jetty",
  memcached: "memcached",
  minio: "minio",
  sonarqube: "sonarqube",
  sentry: "sentry",
};

// Some icons work better with specific variants
const VARIANT_OVERRIDES: Record<string, string> = {
  nodejs: "plain",
  go: "original-wordmark",
  postgresql: "plain",
  redis: "plain",
  mongodb: "plain",
  mysql: "plain",
  nginx: "original",
  python: "plain",
  java: "plain",
  ruby: "plain",
  php: "plain",
  rust: "original",
  docker: "plain",
  kubernetes: "plain",
  grafana: "plain",
  prometheus: "plain",
  elasticsearch: "plain",
  rabbitmq: "original",
  jenkins: "plain",
  apache: "plain",
  ubuntu: "plain",
  debian: "plain",
  alpine: "original",
  react: "original",
  vuejs: "plain",
  angular: "plain",
  svelte: "plain",
  nextjs: "plain",
  spring: "plain",
  django: "plain",
  flask: "original",
  laravel: "original",
  express: "original",
  argocd: "plain",
  gitlab: "plain",
  terraform: "plain",
};

/**
 * Extract technology name from a container image string.
 * Handles formats like:
 *   - "nginx:latest"
 *   - "docker.io/library/nginx:1.25"
 *   - "registry.k8s.io/kube-proxy:v1.28"
 *   - "ghcr.io/org/myapp:sha-abc123"
 *   - "gcr.io/project/image:tag"
 */
function extractImageBase(image: string): string {
  // Remove tag/digest
  let base = image.split("@")[0].split(":")[0];

  // Get last path segment (the actual image name)
  const parts = base.split("/");
  return parts[parts.length - 1].toLowerCase();
}

const CDN_BASE = "https://cdn.jsdelivr.net/gh/devicons/devicon@latest/icons";

const urlCache = new Map<string, string | null>();

export function getContainerIconUrl(image: string): string | null {
  if (!image) return null;
  const base = extractImageBase(image);

  if (urlCache.has(base)) return urlCache.get(base)!;

  // Direct match
  let deviconName = IMAGE_TO_DEVICON[base];

  // Try partial match if no direct hit (only match base inside key, skip short bases to avoid false positives)
  if (!deviconName && base.length > 2) {
    for (const [key, value] of Object.entries(IMAGE_TO_DEVICON)) {
      if (base.includes(key) && key.length > 2) {
        deviconName = value;
        break;
      }
    }
  }

  if (!deviconName) {
    urlCache.set(base, null);
    return null;
  }

  const variant = VARIANT_OVERRIDES[deviconName] ?? "original";
  const url = `${CDN_BASE}/${deviconName}/${deviconName}-${variant}.svg`;
  urlCache.set(base, url);
  return url;
}
